### DISCUSSION TITLE
Variable StringTags %s enormous JavaHeap consumption.

### DISCUSSION THREAD

**Discussion by kech61** [mapsforge/mapsforge#1251] [General]
Hi Emux,

Just solved the problem with variable string tags containing numbers by adding a "#" prefix to those string values.
Now I face the problem that the JavaHeap consumption is up to 3 times compared to V4 maps.

In Detail: I added
ref_cycle=%
ref_hike=%
ref_mtb=%
to seperate the inherited Route-Refs from Relations and the REF of the underlaying way.

This works like a charm - however 25GB of Ram (type=hd!!!) for creating an ordinary map of Austria/UK/etc is far too much.

Do you see any chance to reduce this RAM consumption, maybe by adding a param to limit string size
eg: ref_cycle|hike|mtb=%s7 (7chars max) ??

Best regards
Christian
OpenAndroMaps.org

**Comment by devemux86** [comment]
Hi Christian,

Thanks for the report.

> Just solved the problem with variable string tags containing numbers by adding a "#" prefix to those string values.

Nice solution!
Although someone has to see why it does not work differently.

> Now I face the problem that the JavaHeap consumption is up to 3 times compared to V4 maps.

Have you tested and the memory increase is _only because of the %s variable_?

> Do you see any chance to reduce this RAM consumption, maybe by adding a param to limit string size
eg: ref_cycle|hike|mtb=%s7 (7chars max) ??

What do you suggest, the map-writer cut the strings?
How will this help? The names work well all this time.

I have not played much with %s, only using %f in Mapsforge [tag-mapping](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/config/tag-mapping.xml).

cc @Gustl22

**Comment by kech61** [comment]
Hi Gustl,

Sorry for delayed answer, I'm working hard to move OpenAndroMaps to Win10/WSL.

So I tested the memory consumption of

1.) old V4 writer (an old snapshot I used the last years) 
2.) 0.10 writer with variable vars enabled and string vars %F
3.) 0.10 writer with variable vars disabled

TagMapping is the same for all tests, except the %F for non variable setup.
64GB Ram no swaping.
Java-Heap set to 16GB, type=hd

Map is from Norway, this means lots of LandPolygons, lots of Multipolys, = 1,7GB pbf-File

old V4 snapshot write needs 18hours and runs fine on 12GB Ram
0.10 with varible vars enabled need 21h30min and 16GB RAM at its Limits (crashes at 15GB JHeap-Limit)
0.10 with varible vars disabled need 21h00min and 16GB RAM at its Limits (crashes at 15GB JHeap-Limit)

So the difference for using variable vars on writer 0.10 is in fact negligable.
The difference in demand of resources between V4 and V5 writer ist quite surprising.

The difference in % is much greater on smaler maps, here you can estimate 1/3 longer rendertimes and up to 50% more JavaHeap.
Same machine, same map, same mapping, no swapping.

No complaint - I just write down what I have experineced so far.
For someone who renders some maps for his own thats negligable - for updating 270 Maps its a challenge.

**Comment by Gustl22** [comment]
To be honest, I never had the patience to test big maps with 'hd' mode, as it simply took too long.

You said that 'tag-values' turned off (which results in v4 maps), also took much RAM heap with version 0.10. So the issue might not be related to 'tag-values' or 'v5'.
Can you determine the exact release (date), where the issue was introduced?
Can you also write the exact map-writer command in order to narrow down the possibilities?

As far as I can remember, a lot of optimizations improvements are left to do in code for 'hd' mode, which simply don't apply for 'ram' mode, which therefore always is the better option, when possible.

Wasn't quite busy on this project the last years :/

**Comment by devemux86** [comment]
Reports should be reproducible, so other users can test them and improve the code.
(it is an open source project!)

---

From [documentation](https://github.com/mapsforge/mapsforge/blob/master/resources/tag-mapping.xsd#L44-L49): "%s" is different from "%f" (and more heavy).

You can see map-writer [changelog](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md#changelog) and all changes in its [history](https://github.com/mapsforge/mapsforge/commits/master/mapsforge-map-writer).
